gapis/api/vulkan/transform_read_framebuffer.go

```diff
@@ -51,73 +51,103 @@ type pendingRead struct {
+	return true
+}
+
+func (framebufferTransform *readFramebuffer) SetInnerStateMutationFunction(mutator transform2.StateMutator) {
+	framebufferTransform.stateMutator = mutator
+}
+
+func (framebufferTransform *readFramebuffer) BeginTransform(ctx context.Context, inputCommands []api.Cmd, inputState *api.GlobalState) ([]api.Cmd, error) {
+	framebufferTransform.allocations = NewAllocationTracker(inputState)
+	return inputCommands, nil
+}
+
+func (framebufferTransform *readFramebuffer) EndTransform(ctx context.Context, inputCommands []api.Cmd, inputState *api.GlobalState) ([]api.Cmd, error) {
+	// TODO: Do we need a cmdID here?
+	if err := framebufferTransform.FlushPending(ctx, 0, inputState); err != nil {
+		return inputCommands, err
+	}
+	return inputCommands, nil
+}
+
+func (framebufferTransform *readFramebuffer) ClearTransformResources(ctx context.Context) {
+	framebufferTransform.allocations.FreeAllocations()
+}
+
 // If we are acutally swapping, we really do want to show the image before
 // the framebuffer read.
-func (t *readFramebuffer) Transform(ctx context.Context, id api.CmdID, cmd api.Cmd, out transform.Writer) error {
-	if cmd, ok := cmd.(*InsertionCommand); ok {
-		idx_string := keyFromIndex(cmd.idx)
-		if r, ok := t.injections[idx_string]; ok {
-			// If this command is FOR an EOF command, we want to mutate it, so that
-			// we have the presentation info available.
-			if cmd.callee != nil && cmd.callee.CmdFlags().IsEndOfFrame() {
-				cmd.callee.Mutate(ctx, id, out.State(), nil, nil)
-			}
-			for _, injection := range r {
-				if err := injection.fn(ctx, cmd, injection.res, out); err != nil {
-					return err
+func (framebufferTransform *readFramebuffer) TransformCommand(ctx context.Context, id api.CmdID, inputCommands []api.Cmd, inputState *api.GlobalState) ([]api.Cmd, error) {
+	for _, cmd := range inputCommands {
+		if cmd, ok := cmd.(*InsertionCommand); ok {
+			idxstring := keyFromIndex(cmd.idx)
+			if injectionList, ok := framebufferTransform.injections[idxstring]; ok {
+				// If this command is FOR an EOF command, we want to mutate it, so that
+				// we have the presentation info available.
+
+				if cmd.callee != nil && cmd.callee.CmdFlags().IsEndOfFrame() {
+					cmd.callee.Mutate(ctx, id, inputState, nil, nil)
+				}
+
+				for _, injection := range injectionList {
+					if err := injection.fn(ctx, id, cmd, injection.res, inputState); err != nil {
+						return nil, err
+					}
 				}
+
+				continue
 			}
+		}
 
-			return nil
+		if err := framebufferTransform.writeCommands(id, cmd); err != nil {
+			return nil, err
 		}
-	}
-	if err := out.MutateAndWrite(ctx, id, cmd); err != nil {
-		return err
-	}
-	if len(t.pendingReads) > 0 {
-		if id != api.CmdNoID {
-			return t.FlushPending(ctx, out)
+
+		if _, ok := cmd.(*VkQueueSubmit); ok {
+			if len(framebufferTransform.pendingReads) > 0 {
+				if err := framebufferTransform.FlushPending(ctx, id, inputState); err != nil {
+					return nil, err
+				}
+			}
```

Comment:
There seems to be a change of behavior here, the old code called FlushPending as soon as `id != api.CmdNoID`, the new code waits for a VkQueueSubmit. Is this intended?

Reply:
Yes, because actually the intention of the old code was doing "FlushPending" on the VkQueueSubmit because the only way you can call flushPending while having a pending read is when you have vkQueueSubmit.

This is the scenario: 
   - In command splitter you add an insertion command while you are transforming VkQueueSubmit and you replace it here with an actual command list(see color and depth functions).
   - InsertionCommand you inject creates a pending read and the first command without an id of cmdNoID is the vkQueueSubmit that actually generated that insertion command in command splitter.

I couldn't see if there would be a pending read without a command splitter. But I will double check with Andrew just in case.